test/e2e/storage/testsuites/base.go

```diff
@@ -64,26 +64,37 @@ func getTestNameStr(suite TestSuite, pattern testpatterns.TestPattern) string {
 	return fmt.Sprintf("[Testpattern: %s]%s %s%s", pattern.Name, pattern.FeatureTag, tsInfo.name, tsInfo.featureTag)
 }
 
-// RunTestSuite runs all testpatterns of all testSuites for a driver
+// RunTestSuite defines tests for all testpatterns and all testSuites for a driver
 func RunTestSuite(f *framework.Framework, driver TestDriver, tsInits []func() TestSuite, tunePatternFunc func([]testpatterns.TestPattern) []testpatterns.TestPattern) {
 	for _, testSuiteInit := range tsInits {
 		suite := testSuiteInit()
 		patterns := tunePatternFunc(suite.getTestSuiteInfo().testPatterns)
 
 		for _, pattern := range patterns {
+			if skipUnsupportedTest(suite, driver, pattern) {
+				continue
```

Comment:
Log then skip. The user should know a particular test is detected but skipped due to driver incompatibility.

Reply:
That would make each test suite invocation very verbose, because that information would get printed before running any test, whether the user intends to run the filtered out test or not.

I agree that this is useful information, but we should make it depend on some non-default command line flag. Perhaps we can (ab)use `-ginkgo.dryRun` for this?

Reply:
To further reduce the verbosity, we could print a single log message per test suite and accumulate incompatible drivers in a list. This is still a huge improvement over the verbosity we have today.

I'm not a fan of relying on a non-default flag that could potentially be hard to discover, or in the case of using `-ginkgo.dryRun`, hard to correlate with controlling the log level here.

Reply:
I'm not sure I understand what the output should look like. How many lines should that log message have? What information do you expect?

Ideally, it would be something like:
```
The following tests were not added because they are unsupported:
[sig-storage] In-tree Volumes [Driver: nfs] [Testpattern: Dynamic PV (default fs) block mode] - block mode not supported
...
[sig-storage] CSI Volumes [Driver: gcepd] [Testpattern: Dynamic PV (default fs)] - not running on GCE
...
```

But that will be many lines of output, basically one line for each suite/pattern combination unless we implement some intelligence in RunTestSuite that determines whether all patterns or even all combinations were removed. The other problem is that the reason for skipping is not transported back to RunTestSuite. We either just print the test name prefix and no explanation, or we have to change the return type.

Reply:
I tried to implement something like this, but hit a major snag: there's no way to retrieve the current set of components (i.e. the text from previous `Describe` and `Context` calls) from ginkgo during test registration. The effect is that at the place where an `It` call gets skipped, not enough information is available to record the entire test name. In my current prototype, the inner test name is available and (with some extra effort) the testsuite/pattern part. The driver name could be added, but that's still no guarantee that the resulting string is complete.

Here's what it looks like:
```
$ go test -v ./test/e2e 
W1223 17:21:55.928475   18945 test_context.go:392] Unable to find in-cluster config, using default host : http://127.0.0.1:8080
Dec 23 17:21:55.928: INFO: The --provider flag is not set.  Treating as a conformance test.  Some tests may not be run.
I1223 17:21:56.041325   18945 test_context.go:427] The following tests were not added to the test suite because they are not supported in the current configuration:
??? [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with mount options
??? [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with mount options
??? [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with mount options
??? [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with mount options
??? should create and delete block persistent volumes
??? should create and delete block persistent volumes
??? should create and delete block persistent volumes
??? should create and delete block persistent volumes
??? should create and delete block persistent volumes
```

The prototype commit is here: https://github.com/pohly/kubernetes/commit/20e270a695b13d89a7d13337ab96aa9f0d65eacb

I see two options:
1. take this PR as it is, without logging tests skipped during test registration
2. work with Ginkgo upstream such that it better supports skipped test registration

The latter will take time. Ginkgo is not entirely frozen, but maintenance isn't particularly active either. One of my PRs has been waiting to get merged for a while now.

@verult @msau42: I'll let you guys decide.